src/software/amazon/ionschema/internal/TypeCore.kt

```diff
@@ -0,0 +1,51 @@
+package software.amazon.ionschema.internal
+
+import software.amazon.ion.IonSymbol
+import software.amazon.ion.IonType
+import software.amazon.ion.IonValue
+import software.amazon.ionschema.internal.constraint.ConstraintBase
+import software.amazon.ionschema.Violations
+import software.amazon.ionschema.Violation
+
+/**
+ * Instantiated to represent individual Core Types as defined by the
+ * Ion Schema Specification.
+ */
+internal class TypeCore(
+        name: IonSymbol
```

Comment:
Since the impl is DOM based don't see a problem leaving as is

In general I tend to try and have APIs that are not DOM based but Reader/Writer based to lead customers away from the DOM. This also makes it easier to move off the DOM in the future if necessary, for example to use a new lighter immutable DOM or making the validation code skip scan documents to only materialize values that have constraints associated to them.